src/hooks/index.js

```diff
@@ -337,6 +337,69 @@ exports.restrictToUsersOwnCourses = hook => {
 	});
 };
 
+exports.restrictToUsersOwnLessons = hook => {
+	let userService = hook.app.service('users');
+	return userService.find({
+		query: {
+			_id: hook.params.account.userId,
+			$populate: 'roles'
+		}
+	}).then(res => {
+		let access = false;
+		res.data[0].roles.map(role => {
+			if (role.name === 'administrator' || role.name === 'superhero' )
+				access = true;
+		});
+		if (access)
+			return hook;
+
+		// before-hook
+		if(hook.type === "before") {
+			let populate = hook.params.query.$populate;
+			if (typeof(populate) === 'undefined') {
+				populate = ['courseId'];
+			} else if (Array.isArray(populate) && !populate.includes("courseId")) {
+				populate.push('courseId');
+			}
+			hook.params.query.$populate = populate;
+		} else {
+			// after-hook
+			if (hook.method === "get" && (hook.result||{})._id) {
+				let tempLesson = [hook.result];
+				tempLesson = tempLesson.filter(lesson => {
+					return (hook.params.query.shareToken||{}) === (lesson.shareToken||{}) ||
+					_.some(lesson.courseId.userIds, u => u.toString() === res.data[0]._id.toString()) ||
+					_.some(lesson.courseId.teacherIds, u => u.toString() === res.data[0]._id.toString()) ||
+					_.some(lesson.courseId.substitutionIds, u => u.toString() === res.data[0]._id.toString())
+				});
+				if (tempLesson.length === 0) {
+					throw new errors.Forbidden(`You don't have access to that lesson.`);
+				}
+				hook.result.courseId = hook.result.courseId._id
+			}
+
+			if (hook.method === "find" && ((hook.result||{}).data||[]).length > 0) {
+				hook.result.data = hook.result.data.filter(lesson => {
```

Comment:
Looks like the same in line 369, to use one function will reduce the duplicated code.